src/dataflow/render/reduce.rs

```diff
@@ -64,60 +63,118 @@ where
             let keys_clone = group_key.clone();
             let relation_expr_clone = relation_expr.clone();
 
-            let (ok_input, err_input) = self.collection(input).unwrap();
+            // Our first step is to extract `(key, vals)` from `input`.
+            // We do this carefully, attempting to avoid unneccesary allocations
+            // that would result from cloning rows in input arrangements.
+            let group_key_owned = group_key.to_vec();
+            let aggregates_owned = aggregates.to_vec();
+
+            // Tracks the required number of columns to extract.
+            let mut columns_needed = 0;
+            for key in group_key.iter() {
+                for column in key.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+            for aggr in aggregates.iter() {
+                for column in aggr.expr.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+
+            let mut row_packer = RowPacker::new();
+            let (key_input, mut err_input): (
+                Collection<_, Result<(Row, Row), DataflowError>, _>,
+                _,
+            ) = self
+                .flat_map_ref(input, move |row| {
+                    let temp_storage = RowArena::new();
+                    let mut results = Vec::new();
+                    // First, evaluate the key selector expressions.
+                    // If any error we produce their errors as output and note
+                    // the fact that the key was not correctly produced.
+                    let datums = row.iter().take(columns_needed).collect::<Vec<_>>();
```

Comment:
can we create the vector outside the closure (then clear/fill it from the iter inside) in order to reuse the allocation?

Reply:
We cannot! At least not without a more advanced implementation. Roughly: the lifetime in there is that of `'storage`, which is the lifetime of the `&Row` we get to look at and the lifetime of the associated `Datum` types. That comes in to existence in the `flat_map_ref` call. With some more work (I tried, and I think I walked away from) you can stash a vector like that, and then offer it for use, but it's horrible in the case of "not an arrangement" because there is no such storage and in that case you need to re-alloc a vector for each `Row`..

I'll put a note that we could look in to it though!